### PR TITLE
Real add superblock

### LINKED ISSUE #1543: Fix committee size adjustment during initial synchronization

### LINKED ISSUE #1559: Infinite loop "TargetNotReached: superblock target #{} not reached, requesting more blocks"
In some rare cases when a new node tries to connect to a testnet that been in "no consensus" state for a while, so the testnet has no new blocks, a new node can enter an infinite loop by requesting blocks.

It seems like the node tries to sync up to a target superblock that has already been consolidated, and it is unable to detect that case. A possible solution could be to detect if the target superblock is equal to the last consolidated superblock, and handle that case as "SyncWithCandidate" or "SyncWithoutCandidate".

The attached logs are on commit 967c73f0a46b60dfd93fc0fc726cc47b59b96fff

[2020-09-16-requesting-blocks-loop.zip](https://github.com/witnet/witnet-rust/files/5232790/2020-09-16-requesting-blocks-loop.zip)

### PR DESCRIPTION
**Pull request by lrubiorod**
Closes #1543
Closes #1559

### CODE REVIEW

File: data_structures/src/superblock.rs [hunk 1]
```diff
@@ -387,25 +388,39 @@ impl SuperBlockState {
...
 
-        // Before updating the superblock_beacon, calculate the signing committee
-        self.signing_committee = calculate_superblock_signing_committee(
-            self.ars_previous_identities.clone(),
-            signing_committee_size,
-            superblock_index,
-            self.current_superblock_beacon.hash_prev_block,
-        );
+        let superblock = if let Some(sb) = sync_superblock {
+            // Before updating the superblock_beacon, calculate the signing committee
```
Comment (girazoki) [inline data_structures/src/superblock.rs:398]: Comment about the reason for this

File: data_structures/src/superblock.rs [hunk 2]
```diff
@@ -1311,15 +1351,15 @@ mod tests {
...
+        let sb0 = sbs.build_superblock(&block_headers, ars0, 100, 0, genesis_hash, &alt_keys, None);
         let (v1, v2, v3, v4) = create_votes(sb0.hash(), 0);
         assert_eq!(sbs.add_vote(&v1), AddSuperBlockVote::NotInSigningCommittee);
         assert_eq!(sbs.add_vote(&v2), AddSuperBlockVote::NotInSigningCommittee);
         assert_eq!(sbs.add_vote(&v3), AddSuperBlockVote::NotInSigningCommittee);
         assert_eq!(sbs.add_vote(&v4), AddSuperBlockVote::NotInSigningCommittee);
 
         // The ARS included in superblock 0 contains identities p1, p2, p3
-        let sb1 = sbs.build_superblock(&block_headers, ars1, 100, 1, genesis_hash, &alt_keys);
+        let sb1 = sbs.build_superblock(&block_headers, ars1, 100, 1, genesis_hash, &alt_keys, None);
```
Comment (girazoki) [inline data_structures/src/superblock.rs:1362]: Maybe a test for the case where the sync superblock is not None could help

File: node/src/actors/chain_manager/mod.rs [hunk 3]
```diff
@@ -1279,7 +1292,7 @@ impl ChainManager {
                             };
 
                     // Get the list of members of the ARS with reputation greater than 0
-                    // the list itself is ordered by decreasing reputation
+                        // the list itself is ordered by decreasing reputation
```
Comment (girazoki) [inline node/src/actors/chain_manager/mod.rs:1295]: sorry for this

File: node/src/actors/session/handlers.rs [hunk 4]
```diff
@@ -793,6 +844,12 @@ fn send_inventory_item_msg(session: &mut Session, item: InventoryItem) {
             // Send Transaction msg
             session.send_message(transaction_msg);
         }
+        InventoryItem::SuperBlock(superblock) => {
+            // Build Transaction msg
```
Comment (tmpolaczyk) [inline node/src/actors/session/handlers.rs:848]: Copy-pasted comments

```suggestion
            // Build Superblock msg
```

File: node/src/actors/chain_manager/mod.rs [hunk 5]
```diff
@@ -1207,13 +1212,21 @@ impl ChainManager {
...
+                        log::debug!("Counting votes for a already consolidated superblock index {}", superblock_index);
+                        SuperBlockConsensus::SameAsLocal
+                    }
+                    else if voted_superblock_beacon.checkpoint + 1 != superblock_index {
                         // Warn when there is are missing superblocks between the one that will be
                         // consolidated and the one that will be created
                         log::warn!("Counting votes for Superblock {:?} when the current superblock index is {}", voted_superblock_beacon, superblock_index);
+                        act.chain_state.superblock_state.has_consensus()
+                    }
+                    else{
```
Comment (tmpolaczyk) [inline node/src/actors/chain_manager/mod.rs:1226]: cargo fmt didn't work here

```suggestion
                    } else {
```

File: node/src/actors/chain_manager/handlers.rs [hunk 6]
```diff
@@ -309,6 +311,33 @@ impl Handler<AddBlocks> for ChainManager {
...
+                } else {
+                    log::debug!("Received blocks before superblock");
+                    // Received the `AddBlocks` message before the `AddSuperBlock` message.
+                    // We cannot finish the synchronization without the sync_superblock, so in that
+                    // case, ask for the sync_superblock again, and try to handle the `AddBlocks`
+                    // message later
+                    self.request_sync_target_superblock(ctx, sync_target.superblock);
+                    ctx.notify_later(msg, Duration::from_secs(6));
+                    return;
+                };
```
Comment (tmpolaczyk) [inline node/src/actors/chain_manager/handlers.rs:339]: This logic is a bit hard to understand, I would rewrite it like this: (should be equivalent, but not sure)

```suggestion
                if sync_superblock.is_none() && sync_target.superblock.checkpoint != 0 {
                    log::debug!("Received blocks before superblock");
                    // Received the `AddBlocks` message before the `AddSuperBlock` message.
                    // We cannot finish the synchronization without the sync_superblock, so in that
                    // case, ask for the sync_superblock again, and try to handle the `AddBlocks`
                    // message later
                    self.request_sync_target_superblock(ctx, sync_target.superblock);
                    ctx.notify_later(msg, Duration::from_secs(6));
                    return;
                }
```

File: node/src/actors/session/handlers.rs [hunk 7]
```diff
@@ -845,7 +845,7 @@ fn send_inventory_item_msg(session: &mut Session, item: InventoryItem) {
             session.send_message(transaction_msg);
         }
         InventoryItem::SuperBlock(superblock) => {
-            // Build Transaction msg
+            // Build SuperBlock msg
             let superblock_msg = WitnetMessage::build_superblock(session.magic_number, superblock);
             // Send Transaction msg
```
Comment (tmpolaczyk) [inline node/src/actors/session/handlers.rs:850]: Here too :eyes: 
```suggestion
            // Send SuperBlock msg
```

File: node/src/actors/chain_manager/mod.rs [hunk 8]
```diff
@@ -1207,13 +1212,20 @@ impl ChainManager {
...
+                        SuperBlockConsensus::SameAsLocal
+                    }
+                    else if voted_superblock_beacon.checkpoint + 1 != superblock_index {
                         // Warn when there is are missing superblocks between the one that will be
                         // consolidated and the one that will be created
                         log::warn!("Counting votes for Superblock {:?} when the current superblock index is {}", voted_superblock_beacon, superblock_index);
+                        act.chain_state.superblock_state.has_consensus()
+                    } else {
+                        act.chain_state.superblock_state.has_consensus()
                     }
```
Comment (lrubiorod) [inline node/src/actors/chain_manager/mod.rs:1227]: two branches with same result?

**Review by girazoki — APPROVED** [review]

**Review by tmpolaczyk — APPROVED** [review]

**Pull request merged**